domain/src/main/java/org/oppia/domain/audio/AudioPlayerController.kt

```diff
@@ -0,0 +1,221 @@
+package org.oppia.domain.audio
+
+import android.content.Context
+import android.media.MediaPlayer
+import android.net.Uri
+import androidx.annotation.VisibleForTesting
+import androidx.fragment.app.Fragment
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.MutableLiveData
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+import org.oppia.util.data.AsyncResult
+import org.oppia.util.threading.BackgroundDispatcher
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.locks.ReentrantLock
+import javax.inject.Inject
+import kotlin.concurrent.withLock
+
+/**
+ * Controller which provides audio playing capabilities.
+ * [initializeMediaPlayer] should be used to download a specific audio track.
+ * [releaseMediaPlayer] should be used to clean up the controller's resources.
+ * See documentation for both to understand how to use them correctly.
+ */
+class AudioPlayerController @Inject constructor(
+  private val context: Context,
+  private val fragment: Fragment,
+  @BackgroundDispatcher private val backgroundDispatcher: CoroutineDispatcher
+) {
+
+  inner class AudioMutableLiveData : MutableLiveData<AsyncResult<PlayProgress>>() {
+    override fun onActive() {
+      super.onActive()
+      audioLock.withLock {
+        observerActive = true
+        if (prepared && mediaPlayer.isPlaying)
+          scheduleNextSeekBarUpdate()
+      }
+    }
+
+    override fun onInactive() {
+      super.onInactive()
+      audioLock.withLock {
+        observerActive = false
+        stopUpdatingSeekBar()
+      }
+    }
+  }
+
+  /** Represents current state of internal Media Player. */
+  enum class PlayStatus {
+    PREPARED, // mediaPlayer in "Prepared" state, ready to play(), pause(), seekTo().
+    PLAYING, // mediaPlayer in "Started" state, ready to pause(), seekTo().
+    PAUSED, // mediaPlayer in "Paused" state, ready to play(), seekTo().
+    COMPLETED  // mediaPlayer in "PlaybackCompleted" state, ready to play(), seekTo().
+  }
+
+  /**
+   * [type]: See above.
+   * [position]: Represents mediaPlayer's current position in playback.
+   * [duration]: Represents duration of current audio.
+   */
+  class PlayProgress(val type: PlayStatus, val position: Int, val duration: Int)
+
+  private val mediaPlayer: MediaPlayer by lazy { MediaPlayer() }
+  private var playProgress: AudioMutableLiveData? = null
+  private var nextUpdateJob: Job? = null
+  private val audioLock = ReentrantLock()
+
+  private var prepared = false
+  private var observerActive = false
+  private var mediaPlayerActive = false
+
+  private val SEEKBAR_UPDATE_FREQUENCY = TimeUnit.SECONDS.toMillis(1)
+
+  /**
+   * Loads audio source from a URL and return LiveData to send updates.
+   * This controller cannot already be initialized.
+   */
+  fun initializeMediaPlayer(url: String): LiveData<AsyncResult<PlayProgress>> {
```

Comment:
Don't we also need a method for retrieving the current LiveData? Otherwise during a configuration change, we'd have to release & reinitialize the controller in order to get a new LiveData.